hudi-utilities/src/main/java/org/apache/hudi/utilities/HoodieSnapshotExporter.java

```diff
@@ -0,0 +1,219 @@
+    String basePath = null;
+
+    @Parameter(names = {"--target-base-path", "-tbp"}, description = "Base path for the target output files (snapshots)", required = true)
+    String outputPath = null;
+
+    @Parameter(names = {"--snapshot-prefix", "-sp"}, description = "Snapshot prefix or directory under the target base path in order to segregate different snapshots")
+    String snapshotPrefix;
+
+    @Parameter(names = {"--output-format", "-of"}, description = "e.g. Hudi or Parquet", required = true)
+    String outputFormat;
+
+    @Parameter(names = {"--output-partition-field", "-opf"}, description = "A field to be used by Spark repartitioning")
+    String outputPartitionField;
+  }
+
+  public void export(SparkSession spark, Config cfg) throws IOException {
+    String sourceBasePath = cfg.basePath;
+    String targetBasePath = cfg.outputPath;
+    String snapshotPrefix = cfg.snapshotPrefix;
+    String outputFormat = cfg.outputFormat;
+    String outputPartitionField = cfg.outputPartitionField;
+    JavaSparkContext jsc = new JavaSparkContext(spark.sparkContext());
+    FileSystem fs = FSUtils.getFs(sourceBasePath, jsc.hadoopConfiguration());
+
+    final SerializableConfiguration serConf = new SerializableConfiguration(jsc.hadoopConfiguration());
+    final HoodieTableMetaClient tableMetadata = new HoodieTableMetaClient(fs.getConf(), sourceBasePath);
+    final TableFileSystemView.BaseFileOnlyView fsView = new HoodieTableFileSystemView(tableMetadata,
+        tableMetadata.getActiveTimeline().getCommitsTimeline().filterCompletedInstants());
+    // Get the latest commit
+    Option<HoodieInstant> latestCommit =
+        tableMetadata.getActiveTimeline().getCommitsTimeline().filterCompletedInstants().lastInstant();
+    if (!latestCommit.isPresent()) {
+      LOG.warn("No commits present. Nothing to snapshot");
+      return;
+    }
+    final String latestCommitTimestamp = latestCommit.get().getTimestamp();
+    LOG.info(String.format("Starting to snapshot latest version files which are also no-late-than %s.",
+        latestCommitTimestamp));
+
+    List<String> partitions = FSUtils.getAllPartitionPaths(fs, sourceBasePath, false);
+    if (partitions.size() > 0) {
+      List<String> dataFiles = new ArrayList<>();
+
+      if (!StringUtils.isNullOrEmpty(snapshotPrefix)) {
+        for (String partition : partitions) {
+          if (partition.contains(snapshotPrefix)) {
+            dataFiles.addAll(fsView.getLatestBaseFilesBeforeOrOn(partition, latestCommitTimestamp).map(f -> f.getPath()).collect(Collectors.toList()));
+          }
+        }
+      } else {
+        for (String partition : partitions) {
+          dataFiles.addAll(fsView.getLatestBaseFilesBeforeOrOn(partition, latestCommitTimestamp).map(f -> f.getPath()).collect(Collectors.toList()));
+        }
+      }
+
+      if (!outputFormat.equalsIgnoreCase("hudi")) {
```

Comment:
it would be good to have a `switch` to differentiate the cases and refactor each case to different methods for readability.
Note that though we only have 'hudi' and 'parquet' for now, there could be more cases in future when needs rise.

Reply:
Here, exporter not only supports the conversion of hudi data sets to parquet, but also supports all data types currently supported by spark, such as json or jdbc or avro conversion. The .format () method is the key. Of course, in Exporter, it is only divided into hudi type or not. If it is, then you can copy the relevant files to the specified directory. If not, you can read and convert it. I don't think "switch cases" are needed here.

Reply:
@OpenOpened Understand that the format() can provide some flexibility there but i would argue against it for 2 reasons
1. we are providing this as a user-input argument to establish a contract with value passed to our API. We don't want to rely it on the internal logic where spark takes that and does the conversion. Imagine if user passes an invalid value like "foo" and then Spark API throws an error which will expose the internal implementation which user should not care about.
2. We want to be explicit on what the API supports and does not. `switch` makes it clear and the code more readable. And the `default:` case will be the ideal place to throw "Unsupported" exception to user.
In addition, the `export()` method is sort of lengthy. For better readability, I would suggest break it into multiple methods. (Handling different conversion type could be a good separation)

Reply:
If you use the swich cases code, it might look something like this:
`swich (uotput) {
    case "hudi":exportToHudi();break;
    case "json":exportToJson();break;
    .....
}`

For the implementation part of the method, it is possible to just transform the format() parameters of spark, which is a bit redundant. We can use a compromise solution, using the Spark DataSource lookup method to verify the correctness of the user input format.

Reply:
I uploaded the latest code and gave a "foo" output format test case; -)

Reply:
In addition. DataSource lookup method and can handle the case when the user customizes the DataSrouce implementation.

Reply:
I got your concern. The major point I'd to emphasize here is we want to be explicit on what we support for this utility to the users. 
The spark `format()` API is powerful and reduces the code; on the down side, it makes the code difficult to understand: we claim supporting parquet and hudi, but internally we rely on spark API support. I imagine whatever claimed in the configuration docs should be associated to the code for ease of understanding; that's what I intend the `switch` for. With fall-through switch cases, you can still use `format()` and achieve explicitness.
Though saying that, I don't mind with your current approach and change the Exporter's docs saying support all Spark write format. A little bit worried about spark decoupling in future, but maybe not big issue for now.